### ISSUE TITLE
Black screen when booting Mac build on M1

### ISSUE THREAD

**Issue by dtaddis** [FNA-XNA/FNA#365]
Hi,

I was very excited to start making a Mac app bundle of my XNA game! I followed the [instructions](https://github.com/FNA-XNA/FNA/wiki/3:-Distributing-FNA-Games) like a hawk, and besides having to do some renaming of a few .dylib files (libSDL2-2.0.0.dylib -> libSDL2.dylib for example), it looked like my game was about to boot up... but I just a get a black screen and a beach ball. The music is playing in the background, but no input seems to be triggering anything (I can't hear the menus moving around).

In the terminal window, I can see this message:

2021-10-01 15:11:17.574 Miasma[8033:256626] WARN: Non-standard D32S8 window depth format!
2021-10-01 15:11:17.574 Miasma[8033:256626] INFO: EXT_swap_control_tear unsupported. Fall back to standard VSync.
2021-10-01 15:11:17.574 Miasma[8033:256626] INFO: FNA3D Driver: OpenGL
2021-10-01 15:11:17.574 Miasma[8033:256626] INFO: OpenGL Renderer: Apple M1
2021-10-01 15:11:17.574 Miasma[8033:256626] INFO: OpenGL Driver: 2.1 Metal - 71.7.1
2021-10-01 15:11:17.574 Miasma[8033:256626] INFO: OpenGL Vendor: Apple
2021-10-01 15:11:17.575 Miasma[8033:256626] INFO: MojoShader Profile: glsl120
UNSUPPORTED (log once): POSSIBLE ISSUE: unit 1 GLD_TEXTURE_INDEX_2D is unloadable and bound to sampler type (Float) - using zero texture because texture unloadable

Is this a known error? Have I done something wrong or is it an incompatibility somewhere? I only have an M1 MacBook Air, and I suspect that's the issue, but don't have an Intel Mac to try it against instead.

After a bit of Googling, it feels like an OpenGL bug:

https://developer.apple.com/forums/thread/683865

Could it be possible to use Vulkan and avoid OpenGL entirely?

Thanks,
David

**Comment by flibitijibibo** [comment]
>libSDL2-2.0.0.dylib -> libSDL2.dylib

It sounds like you are using an unsupported runtime. Can I guess that it's CoreCLR?

https://github.com/FNA-XNA/FNA/wiki/0:-FAQ#i-have-a-bug-when-running-on-net-core-and-

You can pass /gldevice:Vulkan to use Metal on M1, but yes this is definitely a Rosetta 2 bug, not FNA.

**Comment by flibitijibibo** [comment]
(This is unlikely but if you are using MonoKickstart make sure FNA.dll.config is in your Resources folder, it's extraordinarily unlikely that your build system didn't already put this next to FNA.dll but who knows.)

**Comment by dtaddis** [comment]
Aha! I was indeed missing FNA.dll.config. I was being very picky what I copied into the bundle and missed that file. Thanks - with that, the original .dylib filenames all work. I was targeting ".NET Framework 4" to match FNA, I'm not using .NET Core (AFAIK).

And forcing it to use Vulkan allows the game to boot! It works great, is there a way to bake that in? Otherwise, I guess I'm waiting for Apple to fix OpenGL support?

**Comment by flibitijibibo** [comment]
You can SetEnvironmentVariable("FNA3D_FORCE_DRIVER", "Vulkan") when SDL_GetPlatform() == "Mac OS X" but I don't know if I would recommend that right away.

One other thing, though I would have hoped Windows/Linux would have lit this up too: Make sure you're not deadlocking with threaded graphics - on your dev environment you can set FNA3D_FORCE_DRIVER=OpenGL. If you've got threaded graphics going on I would just remove that, it'll be faster even on D3D11/Vulkan anyway.

**Comment by dtaddis** [comment]
Wow, thank you so much for the advice. I tried forcing OpenGL on my PC (which usually uses D3D11), and had exactly the same deadlock. It was indeed due to threaded graphics - way back when I was getting it on Xbox, I loaded a level's content on a separate thread while an animated loading screen played.

For now I've just made it load synchronously but it's all working! I still get the 'UNSUPPORTED' message in the terminal, but it looks like it was not the cause of the problem after all. I will probably make it load in slices so I can still animate the loading screen.

Ok, I'm all sorted, I can go and add Mac support in Steamworks now. :)

Thanks,
David
